### PR TITLE
[WIP] Fix MisteryMania - Updated

### PR DESCRIPTION
**Pull request by KiritoDv**
This PR tries to fix an old game called mistery mania, It implements this functions and classes:

 * [Implemented fgetc](https://github.com/hikari-no-yume/touchHLE/commit/6e1ac76b432a46d4da5aaf7bf69f3a5699f66119)
 * [Implemented CGData](https://github.com/hikari-no-yume/touchHLE/commit/734560cf926c34946fe01b15682cddb3eb8a9759)
 * [Implemented NSClass](https://github.com/hikari-no-yume/touchHLE/commit/74d4e449894c668a9b320b537fffe50c3c15bfe4)
 * [Implemented NSBundle::infoDictionary](https://github.com/hikari-no-yume/touchHLE/commit/d12d7e2673841d5a2ecf83f9e317aef2a57383e1)
 * [Implemented CGImageCreateWithPNGDataProvider](https://github.com/hikari-no-yume/touchHLE/commit/1e5d120f7428b482b35750632f9f14b51285a522)
 * [Implemented TexParameteriv, TexSubImage2D, BufferData & Color4ub](https://github.com/hikari-no-yume/touchHLE/commit/b59540bd088ea9ab3c46747ea74d59bd6a35f8ce)
 * [Implemented CopyTexSubImage2D](https://github.com/hikari-no-yume/touchHLE/pull/111/commits/11ddd1c2bb61759f153d7430687708c0bbfc3b2d)

Screenshots: 

image

### CODE REVIEW

**Comment by GamingDuckingalt** [comment]
i don't understand why this PR is closed

File: src/frameworks/opengles/gles_generic.rs [hunk 1]
```diff
@@ -264,4 +277,12 @@ pub trait GLES {
     unsafe fn CheckFramebufferStatusOES(&mut self, target: GLenum) -> GLenum;
     unsafe fn DeleteFramebuffersOES(&mut self, n: GLsizei, framebuffers: *const GLuint);
     unsafe fn DeleteRenderbuffersOES(&mut self, n: GLsizei, renderbuffers: *const GLuint);
+    unsafe fn BufferData(
```
Comment (hikari-no-yume) [inline src/frameworks/opengles/gles_generic.rs:314]: Can you move this to the buffers section? (same in the other files)

File: src/frameworks/opengles/gles1_native.rs [hunk 2]
```diff
@@ -545,4 +577,16 @@ impl GLES for GLES1Native {
...
     }
+    unsafe fn BufferData(
+        &mut self,
+        target: GLenum,
+        size: GLsizeiptr,
+        data: *const GLvoid,
+        usage: GLenum,
+    ) {
+        gles11::BufferData(target, size, data, usage)
+    }
```
Comment (hikari-no-yume) [inline src/frameworks/opengles/gles1_native.rs]: Can you move this to the buffers section? (same in the other files)

File: src/frameworks/opengles/gles1_native.rs [hunk 3]
```diff
@@ -545,4 +577,16 @@ impl GLES for GLES1Native {
...
+    unsafe fn Color4ub(&mut self, red: GLubyte, green: GLubyte, blue: GLubyte, alpha: GLubyte) {
+        gles11::Color4ub(red, green, blue, alpha)
+    }
```
Comment (hikari-no-yume) [inline src/frameworks/opengles/gles1_native.rs]: Can you move this to where the other `Color` functions are? (same in the other files)

File: src/libc/stdio/printf.rs [hunk 4]
```diff
@@ -115,6 +115,18 @@ pub fn printf_inner<const NS_LOG: bool, F: Fn(&Mem, GuestUSize) -> u8>(
                 let description = ns_string::to_rust_string(env, description);
                 write!(&mut res, "{}", description).unwrap();
             }
+            b'x' => {
```
Comment (hikari-no-yume) [inline src/libc/stdio/printf.rs]: Can you put the `%x` part in a separate commit from the `vsnprintf` change?

File: src/image.rs [hunk 5]
```diff
@@ -49,6 +49,19 @@ impl Image {
...
+        let pixels = unsafe {
+            let pixels = std::slice::from_raw_parts_mut(pixels, len as usize);
+            for pixel in pixels.chunks_exact_mut(4) {
+                let [r, g, b, a]: [u8; 4] = pixel.try_into().unwrap();
+                let a = a as f32 / 255.0;
+                pixel[0] = (r as f32 * a).round() as u8;
+                pixel[1] = (g as f32 * a).round() as u8;
+                pixel[2] = (b as f32 * a).round() as u8;
+            }
+            pixels.as_mut_ptr()
```
Comment (hikari-no-yume) [inline src/image.rs:69]: Please also change `CGImageGetAlphaInfo` so it returns the appropriate constant now (and update the comments in it — you can remove the `TODO`).

**Comment by hikari-no-yume** [comment]
I cherry-picked the `fgetc` commit to trunk now (79389dfd0a00df08ceb4f60646ac47aebc488b07).

File: src/frameworks/foundation/ns_class.rs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    class: id, // NSString
+) {
+    // TODO: avoid copy
+    let class_name = ns_string::to_rust_string(env, class);
+
+    log_dbg!("NSClassFromString({:?}, ...)", class_name);
+
+    // TODO: We should use this, but i dont want to make a MPMediaPickerControllerDelegate yet
+    // let class = env.objc.get_known_class(&class_name, &mut env.mem);
+    msg![env; nil new]
```
Comment (hikari-no-yume) [inline src/frameworks/foundation/ns_class.rs:19]: Can you change it to use `get_known_class`, but first have it check if the class is `MPMediaPickerControllerDelegate` and if so return nil? (You can precede it with a comment explaining it's a hack for Mystery Mania and why.)

File: src/frameworks/foundation/ns_bundle.rs [hunk 7]
```diff
@@ -124,7 +124,9 @@ pub const CLASSES: ClassExports = objc_classes! {
        withExtension:(id)extension { // NSString *
    msg![env; this URLForResource:name withExtension:extension subdirectory:nil]
 }
-
+- (id)infoDictionary {
+    nil
+}
```
Comment (hikari-no-yume) [inline src/frameworks/foundation/ns_bundle.rs:155]: If we're going to have this hack, can you add a log message, something like `log!("TODO: [(NSBundle*){:?} infoDictionary] (returning nil)", this);`, so if an app uses this and breaks, there'll be a clue for the problem in the log?

**Review by hikari-no-yume — CHANGES_REQUESTED** [review]
Sorry I took so long to look at this… I have some suggestions.

Note to self: I need to look closer at the CGDataProvider thing at some point.

**Comment by hikari-no-yume** [comment]
I needed the premultiplied alpha thing for something else, and in the process I discovered a bug with your version. Since I wanted other changes made to it anyway, I realised I might as well write my own: 48266f3904987cb4ac30f2085b133f93608cd437. You can drop your commit I guess. Sorry for the wasted effort.

**Comment by hikari-no-yume** [comment]
I ended up having to reimplement a lot of stuff from this pull request in order to save time. Also, I did some work to debug the mysterious audio crashes that required the GDB hacks, and fixed the root cause in e8aceef416e5ed9e19d99adc83793300f6bfdf33. And now (as of d096a8bd305f1503e197f82857ce242c6d2b1617) I'm happy to report that Mystery Mania 1.1.0 is working on trunk!

screenshot of the title screen
in-game screenshot

With that, I can close the PR.

**Pull request closed without merging**